### PR TITLE
Bloom Filters: An Introduction

### PR DESCRIPTION
**Pull request by lalith1403**
I have written an article on bloom filters. There are two images included with the main file. I request the team to kindly review my article.

### CODE REVIEW

File: articles/bloom-filters-data-structure/index.md [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+3. Insert an element into the table
+4. Delete an element from the table
+5. Delete the hash table
+
+### Hash Table
+The hash table is derived from the array. In the case of an array, we have indices that start from either 0 or 1. These indices are the keys and the value stored at these places are the values. Similarly, hash tables work on key-value pairs, where our keys are custom-defined. For example, dictionaries in python are a good example of hash tables.
+
+A hash table is a generalization of the array. With an array, we store the element whose key is k at a position k of the array. That means, given a key k, we find the element whose key is k by just looking in the k position of the array. This is called direct addressing.
+
+The way we index the values determines the position of the value for a given key. If we directly search for the kth element by specifying its index k, again that is called direct addressing. This is feasible in situations when memory is not a constraint. What if we need extra memory? That is, the number of keys is greater than the number of memory locations possible. In such cases, the flexibility of hash tables can be used.
```
Comment (hectorkambow) [inline articles/bloom-filters-data-structure/index.md:49]: Was confused if we are saying the same thing twice :  Might be needed just confused me a bit @lalith1403 
"A hash table is a generalization of the array. With an array, we store the element whose key is k at a position k of the array. That means, given a key k, we find the element whose key is k by just looking in the k position of the array. This is called direct addressing.

The way we index the values determines the position of the value for a given key. If we directly search for the kth element by specifying its index k, again that is called direct addressing. This is feasible in situations when memory is not a constraint. "

File: articles/bloom-filters-data-structure/index.md [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+- Minimal number of collisions
+- Not computationally intensive
+- Uniform distribution of keys for various values
+- Maximum usage of the information present in the values
+
+Let us discuss some of the strategies that complement hash functions and deal with collisions.
+
+#### Separate Chaining
+We use the concept of linked lists to resolve collisions. Upon collision for a given key, the new value is linked to the existing value at the location using linked lists. This method is easy to implement but is memory intensive. Each link adds on a new pointer and therefore additional memory space is required. While using the hash table, if the length of the linked list is too long, then the lookup time increases to O(n) from O(1). This defeats the purpose of implementing a hash table.
```
Comment (hectorkambow) [inline articles/bloom-filters-data-structure/index.md:66]: put a link/source about linked list?
Also ? is separate chaining and linked lists the same thing?

File: articles/bloom-filters-data-structure/index.md [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+excerpt_separator: <!--more-->
+images:
+
+  - url: /engineering-education/bloom-filters-data-structure/hero.jpg
+    alt: Bloom Filters Node.js
+---
+Bloom filters are a probabilistic data structure that uses the concept of hashing extensively. It was designed to solve the problem of finding an element in a set, while keeping in mind high efficiency, memory, and time. In this article, we will be covering the basics of hashing, bloom filters, and the applications of this data structure.
+<!--more-->
+### Hashing
+Hashing is a data structure designed to store and retrieve information in the most efficient manner. It is useful in performing optimal searches and is predominantly used in the construction of symbol tables.
```
Comment (hectorkambow) [inline articles/bloom-filters-data-structure/index.md:20]: put a link to hashing

File: articles/bloom-filters-data-structure/index.md [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+- Uniform distribution of keys for various values
+- Maximum usage of the information present in the values
+
+Let us discuss some of the strategies that complement hash functions and deal with collisions.
+
+#### Separate Chaining
+We use the concept of linked lists to resolve collisions. Upon collision for a given key, the new value is linked to the existing value at the location using linked lists. This method is easy to implement but is memory intensive. Each link adds on a new pointer and therefore additional memory space is required. While using the hash table, if the length of the linked list is too long, then the lookup time increases to O(n) from O(1). This defeats the purpose of implementing a hash table.
+
+#### Open Addressing
+Open addressing is one such technique which stores all the keys in the hash table, rather than extending values via other data structures. The collision is resolved using a process called probing. If a collision occurs, then the next nearest empty location is filled with the value. Such an algorithm is called linear probing. The disadvantage of linear probing is it causes clustering of values if the number of collisions increases.
```
Comment (hectorkambow) [inline articles/bloom-filters-data-structure/index.md:69]: link to open addressing

File: articles/bloom-filters-data-structure/index.md [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+Additionally, what is the advantage of using a bloom filter? Bloom filters reduce the number of calls made to resources such as servers, databases, by quickly eliminating inputs that don't match with the actual value. Let us understand this in a bit more detail, looking at how bloom filters use hashing.
+
+### Explanation
+Designed by Burton Bloom in the year 1970, a Bloom Filter uses a multiple number of hash functions. The bloom filters start with an array, whose bits are initialized to zero. The input is then passed through m number of hash functions. Let the size of the array be n. The number of hash functions, m, is less than the size of the array, n. When the m different hash functions are applied to the value, we get m different values. All these values are set as the bit 1 in the array.
```
Comment (hectorkambow) [inline articles/bloom-filters-data-structure/index.md:79]: need a source here -this is not general knowledge - (anything with a date - needs a source)

File: articles/bloom-filters-data-structure/index.md (hunk 2 above)
Comment (lalith1403) [reply articles/bloom-filters-data-structure/index.md:66]: No... separate chaining uses linked list... have clarified the same... Let me know if it is alright

File: articles/bloom-filters-data-structure/index.md (hunk 1 above)
Comment (lalith1403) [reply articles/bloom-filters-data-structure/index.md:49]: Yep! There was repetition... I have removed it and rephrased the sentence.

**Comment by hectorkambow** [comment]
👋  Hi Kurt @kurt-sectionio - wanted to ask if you could give this _Introduction Bloom Filters_ article a final tech review before we publish - whenever you have a moment. 
Thank you 🚀

**Comment by kurt-sectionio** [comment]
Sorry for the delay. I have not forgotten about this and I will review it but might not get to it until after close of biz today.

**Comment by hectorkambow** [comment]
> Sorry for the delay. I have not forgotten about this and I will review it but might not get to it until after close of biz today.

No worries at all - much appreciated Kurt 🚀

**Review by kurt-sectionio — CHANGES_REQUESTED** [review]
Technically accurate and an interesting topic, however I would like to see a couple revisions to make it more readable:
1. give a brief description of the sock problem so the reader is not required to follow the link;
2. instead of or along with O-notation for complexity, try to describe it in words. The notation is fine, but readers may not know O(2n) => O(n);
3. The section on Hash Table starting at line 44 has redundant paragraphs likening hash table to array data type. 

In general, keep in mind that the ideal reader is intelligent but lacks experience in this particular area and you want to make it easy for them to follow your big idea.

Thank you

**Pull request merged**